Firestore/Swift/Source/Codable/FirestoreDecoder.swift

```diff
@@ -0,0 +1,1189 @@
+        return KeyedDecodingContainer(container)
+    }
+    
+    public func nestedUnkeyedContainer(forKey key: Key) throws -> UnkeyedDecodingContainer {
+        self.decoder.codingPath.append(key)
+        defer { self.decoder.codingPath.removeLast() }
+        
+        guard let value = self.container[key.stringValue] else {
+            throw DecodingError.valueNotFound(UnkeyedDecodingContainer.self,
+                                              DecodingError.Context(codingPath: self.codingPath,
+                                                                    debugDescription: "Cannot get nested unkeyed container -- no value found for key \"\(key.stringValue)\""))
+        }
+        
+        guard let array = value as? [Any] else {
+            let context = DecodingError.Context(codingPath: codingPath, debugDescription: "Not an array")
+            throw DecodingError.typeMismatch([Any].self, context)
+        }
+        
+        return _FirestoreUnkeyedDecodingContainer(referencing: self.decoder, wrapping: array)
+    }
+    
+    private func _superDecoder(forKey key: CodingKey) throws -> Decoder {
+        self.decoder.codingPath.append(key)
+        defer { self.decoder.codingPath.removeLast() }
+        
+        let value: Any = container[key.stringValue] ?? NSNull()
+        return _FirestoreDecoder(referencing: value, at: self.decoder.codingPath)
+    }
+    
+    public func superDecoder() throws -> Decoder {
+        return try _superDecoder(forKey: _FirestoreKey.super)
+    }
+    
+    public func superDecoder(forKey key: Key) throws -> Decoder {
+        return try _superDecoder(forKey: key)
+    }
+}
+
+@available(swift 4.0.0)
+fileprivate struct _FirestoreUnkeyedDecodingContainer : UnkeyedDecodingContainer {
+    // MARK: Properties
+    /// A reference to the decoder we're reading from.
+    private let decoder: _FirestoreDecoder
+    
+    /// A reference to the container we're reading from.
+    private let container: [Any]
+    
+    /// The path of coding keys taken to get to this point in decoding.
+    private(set) public var codingPath: [CodingKey]
+    
+    /// The index of the element we're about to decode.
+    private(set) public var currentIndex: Int
+    
+    // MARK: - Initialization
+    /// Initializes `self` by referencing the given decoder and container.
+    fileprivate init(referencing decoder: _FirestoreDecoder, wrapping container: [Any]) {
+        self.decoder = decoder
+        self.container = container
+        self.codingPath = decoder.codingPath
+        self.currentIndex = 0
+    }
+    
+    // MARK: - UnkeyedDecodingContainer Methods
+    public var count: Int? {
+        return container.count
+    }
+    
+    public var isAtEnd: Bool {
+        return currentIndex >= count!
+    }
+    
+    public mutating func decodeNil() throws -> Bool {
+        guard !isAtEnd else {
```

Comment:
Similar to above, we can factor out a helper for this to cut down on the sheer verbiage required to wade through to read this.

Something like `expectNotAtEnd`?

Reply:
Done